main.cpp

```diff
@@ -60,8 +62,55 @@ struct CommandLineOptions {
     QString updateUpdaterVersion;
     bool updateGame;
     bool playNow;
+    QString connectUrl;
 };
 
+static void argParseError(const QString& message)
+{
+    QMessageBox::critical(nullptr, "Unvanquished Updater", message);
+    exit(1);
+}
+
+QString getURLFromPositionalOptions(const QCommandLineParser& parser)
+{
+    QStringList args = parser.positionalArguments();
+    if (args.empty()) {
+        return "";
+    } else if (args.size() > 1) {
+        argParseError("Too many command line arguments.");
+    }
+
+    QString url = args[0];
+    QRegularExpression scheme("^[a-zA-Z][-+.a-zA-Z]*://");
+    url.replace(scheme, ""); // Strip scheme
+
+    int passwordEnd = url.indexOf('@') + 1; // relies on -1 => not found
+    if (passwordEnd >= url.size()) {
+        argParseError("URL argument is empty");
+    }
+
+    // Square brackets are shell globbing metacharacters but may be used in IPv6 addresses.
+    // Take advantage of a Daemon parsing quirk that it will still be understood without the
+    // leading '['.
+    if (url[passwordEnd] == '[') {
+        url.remove(passwordEnd, 1);
+    }
+
+    // If a custom command is used, we have to consider the pessimistic case
+    // of a command line like   bash -c "%command%"
+    // so metacharacters from any common shells must be forbidden.
+    // What we need to accept are domain names and IP addresses.
+    // Also throw in '@' for limited support of the password feature (unv://hunter2@1.2.3.4)
```

Comment:
Should we not just have an option to shell escape stuff? This seems brittle and hacky.

Reply:
Most of the time it won't be in a shell context, so if it is escaped you won't get the right result.

Another approach could be implementing URL `%` decoding in Daemon for "special" characters. This seems not that different from a shell safety perspective (no longer allowing `]@` but allowing `%`). The advantages of that would be avoiding the `[` erasing hack and allowing any character in the password. Disadvantages would be not working for the current Daemon release and needing to do percent decoding in the updater too for it to forward correctly.

Reply:
Actually that's probably worse since `%` is a shell metacharacter in Windows.